sirepo/pkcli/roles.py

```diff
@@ -5,50 +5,59 @@
 :license: http://www.apache.org/licenses/LICENSE-2.0.html
 """
 from __future__ import absolute_import, division, print_function
+from pykern.pkcollections import PKDict
 from pykern.pkdebug import pkdc, pkdexc, pkdlog, pkdp
 from sirepo.pkcli import admin
 import sirepo.auth
+import sirepo.auth.email
 import sirepo.auth_db
 import sirepo.server
 
 
-def add_roles(uid, *args):  # TODO(e-carlin): roles arg
+def add_roles(uid_or_email, *args):
     """Assign roles to a user
     Args:
-        uid (str): Uid of the user
+        uid_or_email (str): Uid or email of the user
         *args: The roles to assign to the user
     """
-    _check_uid_and_roles(uid, args)
-    sirepo.auth_db.UserRole.add_roles(uid, args)
-    admin.audit_proprietary_lib_files(uid)
+    u = _check_uid_or_email_and_roles(uid_or_email, args)
+    sirepo.auth_db.UserRole.add_roles(u, args)
+    admin.audit_proprietary_lib_files(u)
 
 
 
-def delete_roles(uid, *args):
+def delete_roles(uid_or_email, *args):
     """Remove roles assigned to user
     Args:
-        uid (str): Uid of the user
+        uid_or_email (str): Uid or email of the user
         *args: The roles to delete
     """
-    _check_uid_and_roles(uid, args)
-    sirepo.auth_db.UserRole.delete_roles(uid, args)
-    admin.audit_proprietary_lib_files(uid)
+    u = _check_uid_or_email_and_roles(uid_or_email, args)
+    sirepo.auth_db.UserRole.delete_roles(u, args)
+    admin.audit_proprietary_lib_files(u)
 
 
-def list_roles(uid):
+def list_roles(uid_or_email):
     """List all roles assigned to a user
     Args:
-        uid (str): Uid of the user
+        uid_or_email (str): Uid or email of the user
     """
-    _check_uid_and_roles(uid, [])
-    return sirepo.auth_db.UserRole.search_all_for_column('role', uid=uid)
+    u = _check_uid_or_email_and_roles(uid_or_email, [])
+    return sirepo.auth_db.UserRole.search_all_for_column('role', uid=u)
 
 
-def _check_uid_and_roles(uid, roles):
+def _check_uid_or_email_and_roles(uid_or_email, roles):
     sirepo.server.init()
-    assert sirepo.auth_db.UserRegistration.search_by(uid=uid), \
-        'uid {} not found'.format(uid)
+
+    # POSIT: Uid's are from the base62 charset so an '@' implies an email.
+    if '@' in uid_or_email:
+        u = sirepo.auth.email.AuthEmailUser.search_by(user_name=uid_or_email)
+    else:
+        u = sirepo.auth_db.UserRegistration.search_by(uid=uid_or_email)
+    assert u, \
```

Comment:
This should be a pkcli.command_error, not an assert.

Reply:
https://github.com/radiasoft/sirepo/pull/2515/commits/af8a2bb2a236d342c64f98613d2da0402e268f95